### PR TITLE
Multi-layer piping

### PR DESCRIPTION
**Pull request by ComicIronic**
Pipes and pipe items now have a pipe_layer var which controls which layer they're on.
Current pipes now only connect to other pipes on the same layer as them, and it is possible to stack multiple pipes on a tile.
The layer of a pipe item can be changed by using a pipe planner, aligning with a built pipe using the RPD or a pipe item, or setting the alignment on the pipe dispenser.
Added the isConnectable() proc for OOP way of checking if two pipes can do the do.
Added the layer manifold - this pipe has 5 connectors on one side, and one on the other. It acts as a layer converter.
Pipes change their pixel_x and pixel_y according to their pipe layer.

This gives such wonderful setups as:
### The Twister

Yeah, no
### The Heavy Lifter

Its still a bottleneck
### The "Really, dude?"

Volume and style
### The Octopus

Dont count the legs

### CODE REVIEW

**Comment by IratePirate** [comment]
Absolutely glorious. We can finally stop pipes being forced under walls.

**Comment by FudgePucker** [comment]
Oh boy. The fun you can have with these.
Time to use these as a mommi to make atmos into autistic unpainted spaghetti.

**Comment by FudgePucker** [comment]
Oh man after looking at the twister I think this is FUCKING FANTASTIC. You can still make out complex designs.
You can also have fun antag gimmicks with it.
>one of these valves will put air into this depressursing room, the others will fill it with plasma, good luck.

**Comment by PJB3005** [comment]
:+1: DUDE THIS IS FUCKING AMAZING

**Comment by FudgePucker** [comment]
Now all we need are vigesimal devices for twenty pipes.
Maybe a new atmos based engine to replace the broken rust engine.

**Comment by SkowronX** [comment]
It begins. Now we see what bugs need ti hunted down.

**Comment by Exxion** [comment]
Holy shit I was thinking about exactly this a few weeks ago except my idea for how to select the layer was vastly worse
:+1::heavy_multiplication_x::100::arrow_up::arrow_up::100:

Is there a way to connect two pipes on different layers other than with the manifold? I think I had a reason for wanting this ability, but I don't remember what it was.

**Comment by ComicIronic** [comment]
>Is there a way to connect two pipes on different layers other than with the manifold?

No.
I was originally working on something like that, but I decided it was redundant thanks to the manifold anyways.

**Comment by BestName** [comment]
So how does this affect vent crawling?
How do you choose which pipe to enter?

**Comment by d3athrow** [comment]
This disturbs me

**Comment by FudgePucker** [comment]
What? You don't like autistic spaghetti?
This will make optimizing atmos so much easier.

**Comment by Anonus** [comment]
This is such an atrocity, I love it.

**Comment by ComicIronic** [comment]
>ventcrawling

I knew I'd forgotten something.
I'll push a fix for that and meters in a bit.

**Comment by xxalpha** [comment]
This is unquestionably the best feature.

**Comment by DrCelt** [comment]
Beautiful.

**Comment by FudgePucker** [comment]
I can't wait to make pipe mazes for the other mommis to solve

**Comment by SangriaMonkey** [comment]
Can you stick h/e pipes in those?

**Comment by FudgePucker** [comment]
That might be too op.
You could put 5 h/e pipes in the same place

**Comment by SangriaMonkey** [comment]
But that'd be sweet.
Or five lines of manifolds with thermal plates. Seven kelvin or bust.

**Comment by dylanstrategie** [comment]
So wait, how do you choose a pipe's layer when wrenching it down ?

Also, I'll likely attempt the general re-piping once this is released, given that I already re-piped the entire map back when half of the map's piping when under walls

**Comment by SangriaMonkey** [comment]
I wouldn't redo the whole station. Having to alt click constantly to see/repair disposal pipes would be such a fucking pain.

**Comment by FudgePucker** [comment]
I don't think that having unary devices was going to be allowed.
20 vents, scrubbers, injectors, and thermal plates.
Although the thought does get me ROCK HARD

**Comment by VelardAmakar** [comment]
Holy fucking shit.

**Comment by ComicIronic** [comment]
>So wait, how do you choose a pipe's layer when wrenching it down ?

The pipe dispenser machine dispenses pipes at a set layer.
The RPD can dispense pipes at the same layer as a pipe that it is used on.
You can use a pipe on a pipe planner to align in to the desired layer.
You can use a pipe on an existing pipe to align it to that pipe's layer.

**Comment by Exxion** [comment]
The overlap issue with the far left/top pipe layer and the manifold irks me

**Comment by Exxion** [comment]
And bottom/far right on the other end

Where the fuck is the edit comment button on mobile?
inb4 >Using mobile

**Comment by FudgePucker** [comment]
Then my twenty injectors and scrubbers will drive you up the wall

**Comment by ComicIronic** [comment]
Alright, meters and ventcrawling are now up-to-date as well.

**Comment by FudgePucker** [comment]
Noice

**Comment by AutismSpider** [comment]
oh good this will make it easier to hook the TEG to Atmos same with super matter shard room to Atmos

**Comment by demonfiren** [comment]
I can atmos so hard to this.

**Comment by Aranclanos** [comment]
MULTI
TRACKING
PIPING!?

DEJA VU I'VE JUST BEEN ON THIS PLACE BEFORE
HIGHER ON THE STREET AND I KNOW ITS MY PLACE TO GO
WOOOOOOOOOOOOOOO

**Comment by ririchiyo** [comment]
Does this break H/E pipes that can already stack on the same tile up to three times?

**Comment by ComicIronic** [comment]
What do you mean?

**Comment by ririchiyo** [comment]
Forgive the terrible webm I threw together in seconds: https://λ.pw/y7x

**Comment by Exxion** [comment]
I hope so because that was horrifying
I especially hope so because if it doesn't, it would presumably allow for 15 HE pipes on the same tile

**Comment by FudgePucker** [comment]
I want to be able to make octopuses and other pipe designs everywhere on the station as a MoMMI :3
Merge when?

**Comment by d3athrow** [comment]
THIS IS 100% DISGUSTING DO PEOPLE REALLY WANT THIS MERGED?

**Comment by Probe1** [comment]
Eh.  I'd love to have multiple pipes per tile.  We come up with some creative, and pitiful, designs to get around it.

**Comment by PJB3005** [comment]
@d3athrow, is that even a question? MERGE NOW!

**Comment by FudgePucker** [comment]
@d3athrow 
YOU CAN HAVE TWENTY THERMAL PLATES ON THE SAME TILE
TWENTY
TWWWWWEEEEEEEEEEENNNNNNNNNNNTTTTTTTTTTYYYYYYYYYYYYYYYYYYYYYYYYY
JUST LOOK AT THAT HEAVY LIFTER IMAGE, FUCKIN LOOK AT THAT SHIT, OH GOD I AM SO HARD RIGHT NOW HRRRRRRRRRRRNNNNNNNNGGGGGGGGGGGGGG.



Merge when?

**Comment by clusterfack** [comment]
Devices, excluding like pumps/valves, shouldn't be made stackable beyond the already existing stackability in 4 directions

**Comment by PJB3005** [comment]
If you're limiting stacking 4 devices to a tile, atleast make some devices such as vents capable of connecting with multiple channels of pipes

**Comment by ComicIronic** [comment]
>make some devices such as vents capable of connecting with multiple channels of pipes

I don't understand this.
If you mean being able to change the layer of vents, you can do that.
If you mean being able to connect a vent to more than one layer, use a manifold dummy.

**Comment by PJB3005** [comment]
Well, nevermind that actually, I was dumb.

**Comment by FudgePucker** [comment]
@clusterfack 
Deep down inside your heart, there is some part of you screaming out, 'YES, YES, TWENTY INJECTORS/SCRUBBERS/THERMAL PLATES ON THE SAME SPOT. BALANCE CAN GO FUCK ITSELF.'

**Comment by clusterfack** [comment]
No that part of my mind is saying "OH GOD IF I WANT TO BE OPTIMAL ANY BUILD I MAKE WILL TAKE FIVE TIMES LONGER"

**Comment by FudgePucker** [comment]
@clusterfack 
Ah didn't think of it that way.

**Comment by Boxe** [comment]
dark magic

**Comment by FudgePucker** [comment]
https://github.com/d3athrow/vgstation13/issues/5163
I came up with an idea that would use this.

**Comment by Kurfursten** [comment]
When?

**Comment by Boxe** [comment]
never

**Comment by Duny-** [comment]
This is terrifying. I was all for distro and waste on the same tile so we could finally get rid of those pipes under rwalls, but this... This is too much, this is going too far.

**Comment by ComicIronic** [comment]
Don't open Pandora's box unless you can handle what's inside.

**Comment by PJB3005** [comment]
inb4 I will need to maintain this.

comic pls

**Comment by SkowronX** [comment]
# W H E N
# H
# E
# N

File: code/__HELPERS/maths.dm [hunk 1]
```diff
@@ -14,26 +14,22 @@ var/list/sqrtTable = list(1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4,
...
-	//writepanic("[__FILE__].[__LINE__] (no type)([usr ? usr.ckey : ""])  \\/proc/arctan() called tick#: [world.time]")
+proc/arctan(x)
 	var/y=arcsin(x/sqrt(1+x*x))
 	return y
 
-/proc/Ceiling(x, y = 1)
-	//writepanic("[__FILE__].[__LINE__] (no type)([usr ? usr.ckey : ""])  \\/proc/Ceiling() called tick#: [world.time]")
-	. = -round(-x / y) * y
+/proc/Ceiling(x)
+	return -round(-x)
```
Comment (PJB3005) [inline code/__HELPERS/maths.dm]: Simply comical

**Comment by Probe1** [comment]
comics da best

**Comment by Aranclanos** [comment]
VG CUCKS
TG MASTERRACE

**Comment by PJB3005** [comment]
@Aranclanos oi nerd we merged this on GitLab, come at us.

**Comment by Aranclanos** [comment]
gitlab more like git GOODBYE

**Comment by clusterfack** [comment]
TG blown the fuck out

**Pull request closed without merging**